---
title: "splat_sim"
author: "Cody Heiser"
date: "2/26/2020"
output: html_document
---
C. Heiser, 2020

__The purpose of this notebook is to perform single-cell data simulations for testing with dimensionality reduction structural preservation framework.__

We are using the [Splatter](https://github.com/Oshlack/splatter) framework to simulate scRNA-seq datasets with discrete and continuous global structures.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Importing packages

```{r}
library(splatter)
```

## Building parameters for simulating dataset

```{r}
params.path <- newSplatParams(nGenes=500, batchCells=3060, lib.loc=10, lib.scale=0.05, path.nSteps=1000, group.prob=c(0.3333333,0.3333333,0.3333334), path.from=c(0,0,0), de.prob=0.5)#, lib.loc=8, lib.scale=0.2)
sim.paths <- splatSimulatePaths(params.path)
```

```{r}
normalize(sim.paths) -> sim.paths
plotPCA(sim.paths, colour_by="Step")
```

```{r}
plotPCA(sim.paths, colour_by="Group")
```

```{r}
counts(sim.paths)[1:10,1:10]
```

```{r}
setwd("~/git/DR-structure-preservation")
write.csv(counts(sim.paths), file = "inputs/sim_path.csv")
write.csv(colData(sim.paths), file = "inputs/sim_path_obs.csv")
```

## Now apply dimension reduction methods to simulation
```{r}
setwd('~/git/scrna2019/') # insert relative path to GLM-PCA directory
source('algs/glmpca.R')
setwd('~/git/DR-structure-preservation/dev/')
source('fcc_utils.r')
library(zinbwave, quietly = T)
#library(scRNAseq, quietly = T)
#library(SIMLR, quietly = T)
```

***
### ZINB-WaVE
From [Risso, _et al._ (2018)](https://www.nature.com/articles/s41467-017-02554-5), ZINB-WaVE is a zero-inflated negative binomial model for low-dimensional representations of scRNA-seq counts data. Available on [Bioconductor](https://bioconductor.org/packages/release/bioc/html/zinbwave.html).

Create SummarizedExperiment object from counts data, then perform ZINB-WaVE.
```{r}
# transpose matrix to genes x cells format
#sim_expt <- SummarizedExperiment(assays = list(counts=data.matrix(t(colon))))
sim.paths.select <- sim.paths[rowSums(counts(sim.paths))>0,]
# perform ZINB-WAVE analysis
colon_zinbwave <- zinbwave(sim.paths.select, K=2, epsilon=1000, verbose=T)
```

Plot resulting 2D embedding:
```{r}
plot.DR(data.frame(reducedDim(colon_zinbwave)), colorby = sim.paths$Step, name = 'ZINB-WAVE')
```

Output ZINB-WaVE results to .csv files
```{r}
write.csv(data.frame(colon_zinbwave@reducedDims$zinbwave), file = 'outputs/colon_ZINB-WAVE.csv', row.names = F)
```

***
### GLM-PCA
From [Townes, _et al._ (2019)](https://www.biorxiv.org/content/10.1101/574574v1), GLM-PCA is a generalized linear model-based PCA for dimensionality reduction of scRNA-seq counts data. Available on [GitHub](https://github.com/willtownes/scrna2019).

Let's start with colon dataset again
```{r}
# perform GLM-PCA analysis
colon_glmpca <- glmpca(Y=counts(sim.paths.select), L=2, verbose=T)
# plot results
plot.DR(colon_glmpca$factors, colorby = sim.paths$Step, name='GLM-PCA')
```

Output GLM-PCA results to .csv files
```{r}
write.csv(colon_glmpca$factors, file = 'outputs/colon_GLM-PCA.csv', row.names = F) # colon data
```

***
### SIMLR
From [Wang, _et al._ (2017)](https://www.ncbi.nlm.nih.gov/pubmed/28263960), SIMLR is a multikernel learning approach to dimensionality reduction, clustering, and visualization of scRNA-seq counts data. Available on [Bioconductor](https://bioconductor.org/packages/release/bioc/html/SIMLR.html).

First on the colon data.
We can normalize our counts data and feed it into the SIMLR algorithm with the expected number of clusters from Seurat.
```{r}
colon_norm <- arcsinh.norm(colon, margin=1) # normalize by arcsinh-tranforming fractional counts per gene

# perform SIMLR analysis with the estimated number of clusters from prior analysis of data
colon_SIMLR <- SIMLR(t(colon_norm), c = 6, no.dim = 2, normalize = F)
```

```{r}
# plot results
plot.DR(data.frame(colon_SIMLR$ydata), colorby = colon.clu$V1, name='SIMLR')
```

```{r}
# plot results colored by SIMLR's clusters
plot.DR(data.frame(colon_SIMLR$ydata), colorby = colon_SIMLR$y$cluster, name='SIMLR')
```

Perform same workflow on retina data
```{r}
retina_norm <- arcsinh.norm(retina, margin=1) # normalize by arcsinh-tranforming fractional counts per gene

# perform SIMLR analysis with the estimated number of clusters from prior analysis of data
retina_SIMLR <- SIMLR(t(retina_norm), c = 9, no.dim = 2, normalize = F)
```

```{r}
# plot results
plot.DR(data.frame(retina_SIMLR$ydata), colorby = retina.clu$V1, name='SIMLR')
```

```{r}
# plot results colored by SIMLR's clusters
plot.DR(data.frame(retina_SIMLR$ydata), colorby = retina_SIMLR$y$cluster, name='SIMLR')
```

Finally write outputs for 10-D latent space and 2D projections from SIMLR
```{r}
write.csv(colon_SIMLR$F, file = 'outputs/colon_SIMLR_F.csv', row.names = F)
write.csv(colon_SIMLR$ydata, file = 'outputs/colon_SIMLR_ydata.csv', row.names = F)

write.csv(retina_SIMLR$F, file = 'outputs/retina_SIMLR_F.csv', row.names = F)
write.csv(retina_SIMLR$ydata, file = 'outputs/retina_SIMLR_ydata.csv', row.names = F)
```
